Fix: Ignore incorrect hashes and save new hash to cache

german1608 opened the pull request:
Fixes #1790 

```bash
# First time download
$ echo 'https://prelude.dhall-lang.org/v16.0.0/JSON/Type sha256:40edbc9371979426df63e064333b02689b969c4cfbbccfa481216d2d1a6e9759' | dhall
∀(JSON : Type) →
∀ ( json
  : { array : List JSON → JSON
    , bool : Bool → JSON
    , double : Double → JSON
    , integer : Integer → JSON
    , null : JSON
    , object : List { mapKey : Text, mapValue : JSON } → JSON
    , string : Text → JSON
    }
  ) →
  JSON

# Corrupting cache
$ echo 'bla' > ~/.cache/dhall/122040edbc9371979426df63e064333b02689b969c4cfbbccfa481216d2d1a6e9759

# Second time download detects hashes changes and raises a warning
$ echo 'https://prelude.dhall-lang.org/v16.0.0/JSON/Type sha256:40edbc9371979426df63e064333b02689b969c4cfbbccfa481216d2d1a6e9759' | dhall

Warning: Import integrity check failed

Expected hash:

↳ 40edbc9371979426df63e064333b02689b969c4cfbbccfa481216d2d1a6e9759

Actual hash:

↳ 00e3261a6e0d79c329445acd540fb2b07187a0dcf6017065c8814010283ac67f

Interpreter has automatically fixed cached import
∀(JSON : Type) →
∀ ( json
  : { array : List JSON → JSON
    , bool : Bool → JSON
    , double : Double → JSON
    , integer : Integer → JSON
    , null : JSON
    , object : List { mapKey : Text, mapValue : JSON } → JSON
    , string : Text → JSON
    }
  ) →
  JSON
```
What I don't know is how to write tests for this yet

sjakobi commented:
The test I'm referencing in https://github.com/dhall-lang/dhall-haskell/issues/1791 should cover this.

I believe we just need to update this test scaffolding

https://github.com/dhall-lang/dhall-haskell/blob/b0ac16f124b791154de2b17eab63d38a2cbcc0cc/dhall/tests/Dhall/Test/Import.hs#L75-L94

It would be nice if we could actually follow [the upstream instructions](https://github.com/dhall-lang/dhall-lang/tree/master/tests#running-import-tests) more closely here instead of conditionally setting the `XDG_CACHE_HOME` env var, but there might be some non-obvious trickiness involved…

BTW, it's a bit difficult to review PRs with so many style changes. Maybe you could commit those separately so it's easier to find the actual changes to the program logic.

german1608 commented:
> BTW, it's a bit difficult to review PRs with so many style changes. Maybe you could commit those separately so it's easier to find the actual changes to the program logic.
Yes sorry! I tend to fix those kind of issues on the way but I agree, that adds a lot of noise to PRs. Should I change this PR or I can left it as it currently is?

german1608 commented:
I'm gonna change those and set this PR for ready for review

Gabriel439 commented on dhall/src/Dhall/Import.hs:
```diff
@@ -506,20 +504,22 @@ loadImportWithSemanticCache
-                Right e   -> return e
+                    return (ImportSemantics {..})
+                else do
+                    printWarning $ makeHashMismatchMessage semanticHash actualHash
```
Perhaps extend the warning message to say that the interpreter has automatically fixed the cached import

german1608 replied:
I updated PR description with new message

Gabriel439 commented on dhall/src/Dhall/Import.hs:
```diff
@@ -506,20 +504,22 @@ loadImportWithSemanticCache
+                    return (ImportSemantics {..})
+                else do
+                    printWarning $ makeHashMismatchMessage semanticHash actualHash
+                    saveCacheToFile -- TODO: here we should show a warning that caches doesn't match
```
Is this TODO fixed now?

Gabriel439 commented on dhall/src/Dhall/Import.hs:
```diff
@@ -1032,6 +1029,15 @@ loadWith expr₀ = case expr₀ of
 load :: Expr Src Import -> IO (Expr Src Void)
 load = loadRelativeTo "." UseSemanticCache
 
+printWarning :: (MonadCatch m, Alternative m, MonadIO m) => String -> m ()
```
I believe the `MonadCatch` and `Alternative` constraints are not necessary

Gabriel439 commented on dhall/src/Dhall/Import.hs:
```diff
@@ -510,20 +526,25 @@ loadImportWithSemanticCache
+                    printWarning $
+                        makeHashMismatchMessage semanticHash actualHash
+                        <> "\n"
+                        <> "Interpreter has automatically fixed cached import"
```
```suggestion
                        <> "The interpreter has automatically fixed the cached import\n"
```

german1608 commented:
I added the test as @Gabriel439 suggested in [this](https://github.com/dhall-lang/dhall-haskell/pull/1793#pullrequestreview-412290167) comment without actually understanding how that code works. Not the way I'd like to, but prefer to spend time understanding how it works while you review the PR.

@sjakobi I fixed imports style but not other fixes as remove redundants "do" or parenthesis removes because there were so many and I got messy doing so, but I'll try on my next non-style PRs to avoid doing so.

Gabriel439 commented:
@german1608: Whoops, I got confused linked to the wrong place to add the test.  The place to actually add the test is here:

https://github.com/dhall-lang/dhall-haskell/blob/9370394cb045d02d6c9f6960337cdb22a0089d20/dhall/tests/Dhall/Test/Import.hs#L83-L85

sjakobi commented on dhall/src/Dhall/Import.hs:
```diff
@@ -510,20 +526,25 @@ loadImportWithSemanticCache
-        Nothing -> do
+        Nothing -> saveCacheToFile
+    where
+        saveCacheToFile = do
```
There's a lot more going on here than just writing a new cache file. Not sure how to best capture all that in a name. Maybe just use something generic like `fetch`?

sjakobi commented on dhall/src/Dhall/Import.hs:
```diff
@@ -510,20 +526,25 @@ loadImportWithSemanticCache
+                    printWarning $
+                        makeHashMismatchMessage semanticHash actualHash
+                        <> "\n"
+                        <> "The interpreter has automatically fixed cached import\n"
```
We can't really promise that, we can only try. Not sure how to put that into a message…

Gabriel439 replied:
```suggestion
                        <> "The interpreter will attempt to fix the cached import\n"
```

german1608 commented:
@Gabriel439 don't worry! I'll update it

german1608 commented:
I think this is ready for review

german1608 commented:
@sjakobi will this PR fix #1791 ?

sjakobi commented:
> @sjakobi will this PR fix #1791 ?

Hopefully! :)

Does running the `IgnorePoisoned` test dirty the git repo though? If so, we should give the [recommendation](https://github.com/dhall-lang/dhall-lang/tree/master/tests#running-import-tests) to treat it as read-only a try. Or alternatively use a temporary directory for the cache!?

german1608 commented:
> > @sjakobi will this PR fix #1791 ?
> 
> Hopefully! :)
> 
> Does running the `IgnorePoisoned` test dirty the git repo though? If so, we should give the [recommendation](https://github.com/dhall-lang/dhall-lang/tree/master/tests#running-import-tests) to treat it as read-only a try. Or alternatively use a temporary directory for the cache!?

I forgot to check that. Let me see

german1608 commented:
@sjakobi there was a bug when checking which files to set and unset cache env variables.

Looks like `Turtle.filename` returns the filename of a path, so `unit/asLocation/HashA.dhall` and (new) `unit/IgnorePoisonedCacheA.dhall` checks didn't do anything since the filename of them are just `HashA.dhall` and `unit/IgnorePoisonedCacheA.dhall`.

I've fixed that bug using `Turtle.match` and `Turtle.ends` and yes, a new file is created on the cache. I'll try to create a temp dir for each cache run copying from the cache on the dhall-lang submodule

german1608 commented on dhall/tests/Dhall/Test/Import.hs:
```diff
@@ -70,22 +70,32 @@ successTest path = do
-                     ]
-                    then do
-                        setCache
-                        _ <- load
+        let usesCache = [ "hashFromCacheA.dhall"
+                        , "unit/asLocation/HashA.dhall"
+                        , "unit/IgnorePoisonedCacheA.dhall"]
+
+        let endsIn path' = not $ null $ Turtle.match (Turtle.ends path') path
+
+        let buildNewCache = do
+                                tempdir <- Turtle.mktempdir "/tmp" "dhall-cache"
+                                Turtle.liftIO $ Turtle.cptree originalCache tempdir
+                                return tempdir
+
+        let runTest =
+                if any endsIn usesCache
+                    then Turtle.runManaged $ do
+                        cacheDir <- buildNewCache
+                        setCache $ Turtle.format Turtle.fp cacheDir
+                        _ <- Turtle.liftIO load
```
Take a look at this please. This is how I managed to follow [import tests suggestion](https://github.com/dhall-lang/dhall-lang/tree/master/tests#running-import-tests) from dhall-lang repo. I created a temp dir for each file test that may write to the cache

Gabriel439 commented on dhall/tests/Dhall/Test/Import.hs:
```diff
@@ -70,22 +70,32 @@ successTest path = do
+        let endsIn path' = not $ null $ Turtle.match (Turtle.ends path') path
+
+        let buildNewCache = do
+                                tempdir <- Turtle.mktempdir "/tmp" "dhall-cache"
```
It's slightly better if this use the `TMPDIR` environment variable if set (falling back to `"/tmp"` if unset), in order to interoperate better with Nix.  The reason why is that Nix sets `TMPDIR` during a build so that the build doesn't need to write outside of the temporary build directory.
